### ISSUE TITLE
[Plan] 集成可配置的 Muon + AdamW 混合优化器

### ISSUE THREAD

**Issue by Puiching-Memory** [Puiching-Memory/TAAC_2026#53]
# 集成可配置的 Muon + AdamW 混合优化器

更新时间：2026-04-26

## 背景

当前共享 PCVR 训练链路的优化器逻辑集中在 `src/taac2026/infrastructure/pcvr/trainer.py::PCVRPointwiseTrainer`，但实现是硬编码的：

- 模型暴露 `get_sparse_params()` 时，embedding 参数走 `torch.optim.Adagrad`。
- 非 embedding 参数统一走 `torch.optim.AdamW(lr=lr, betas=(0.9, 0.98))`。
- 如果没有 sparse 参数，则全部参数走 AdamW。
- `reinit_high_cardinality_params()` 后 sparse optimizer 会被硬编码重建为 Adagrad。
- `parse_pcvr_train_args` 只暴露 `--lr`、`--sparse_lr`、`--sparse_weight_decay`，没有 optimizer 类型选择，也没有 Muon 参数。

仓库文档中已经多次提到 Muon + AdamW 混合优化器：TAAC 2025 冠军方案使用 Muon + AdamW，UniRec 文档也写到“Embedding 走 RowWiseAdagrad，2D 结构化权重走 Muon，其余参数走 AdamW”。但当前代码还没有对应实现。

旧的 #21 大重构计划中也提到了 Muon，不过它已经关闭，且范围绑定 TorchRec/fbgemm/Triton 大迁移。这个 issue 聚焦当前 PCVR runtime：先在现有 `PCVRParquetDataset + PCVRPointwiseTrainer` 链路上加入可配置优化器栈，保持默认行为不变，再给 Muon 实验留出安全入口。

## 目标

引入一个灵活、可复现、默认不改变现有行为的 optimizer builder，使训练可以在以下策略间切换：

1. **legacy 默认策略**：保持当前行为，embedding/稀疏参数用 Adagrad，dense 参数用 AdamW。
2. **adamw 策略**：所有非 sparse 参数走 AdamW，可显式控制 betas、eps、weight decay。
3. **muon_adamw 混合策略**：embedding/稀疏参数继续走 Adagrad；适合 Muon 的 2D 结构化权重走 Muon；bias、norm、embedding、小型或不适配参数走 AdamW。

核心要求是灵活性：Muon 不应被写死到所有 dense 参数上，也不应迫使每个 `config/*` 实验包自定义 trainer。

### 1. 把 optimizer 构建从 trainer 中抽出来

建议新增：

- `src/taac2026/infrastructure/pcvr/optimizers.py`

职责：

- 定义 optimizer 配置对象，例如 `PCVROptimizerConfig`。
- 定义参数分组逻辑，例如 `split_pcvr_optimizer_params(model, config)`。
- 定义 builder，例如 `build_pcvr_optimizers(model, config)`。
- 返回命名 optimizer 列表或轻量 wrapper，供 trainer 统一 `zero_grad()` / `step()`。

`PCVRPointwiseTrainer` 应只保留训练循环，不再直接决定 AdamW / Adagrad / Muon 的具体构造。

### 2. 保持默认行为完全兼容

默认配置建议是：

```text
--optimizer legacy
--sparse_optimizer adagrad
--dense_optimizer adamw
```

在默认参数下，训练日志、参数分组和更新语义应与当前版本一致：

- 有 `get_sparse_params()`：sparse Adagrad + dense AdamW。
- 无 sparse 参数：AdamW all params。
- sparse reinit 后仍然重建 sparse optimizer，并保留低基数 embedding 的 optimizer state。

### 3. Muon 只作用于合适的结构化参数

Muon 的目标应是 2D 或更高维的结构化 dense weight，而不是所有参数。

建议默认选择规则：

- **进入 Muon**：非 embedding、非 sparse、`parameter.ndim >= 2`、浮点 dtype、`requires_grad=True` 的权重，例如 `nn.Linear.weight`、attention projection、FFN projection。
- **留给 AdamW**：bias、一维 norm 权重、标量参数、embedding 参数、显式排除的模块/参数。
- **留给 sparse optimizer**：`EmbeddingParameterMixin.get_sparse_params()` 返回的 embedding 参数。

需要提供覆盖口，例如：

- `--muon_min_ndim 2`
- `--muon_include_patterns` / `--muon_exclude_patterns`，可选，先支持逗号分隔的参数名 substring 即可。
- `--muon_fallback_optimizer adamw`

### 4. Muon 实现策略

第一阶段建议采用仓库内纯 PyTorch 实现，不新增强依赖：

- 新增一个 `Muon(torch.optim.Optimizer)` 实现。
- 使用 Newton-Schulz 迭代做矩阵正交化，内部计算转为 fp32，更新后回写原 dtype。
- 对非 2D 参数直接拒绝进入 Muon group，避免静默错误。
- CPU 可跑单元测试；CUDA/GPU 性能优化可以后续再做。

可暴露的 Muon 参数：

- `--muon_lr`，默认可先复用 `--lr`。
- `--muon_momentum`。
- `--muon_nesterov`。
- `--muon_ns_steps`，Newton-Schulz 迭代次数。
- `--muon_weight_decay`。
- `--muon_update_scale` 或 `--muon_normalize_update`，用于控制更新幅度。

如果后续要接 Triton/fused Muon，可以在同一 builder 下新增实现后端：

```text
--muon_backend torch|triton
```

第一阶段只需要 `torch`。

### 5. Trainer 训练循环改造

当前 trainer 有 `self.dense_optimizer` 和 `self.sparse_optimizer` 两个字段。为后续灵活性，建议改为：

- `self.optimizers`: list of named optimizer records。
- 每个 record 至少包含：`name`、`kind`、`optimizer`、`param_count`。
- `_train_step` 中循环调用所有 optimizer 的 `zero_grad()` 和 `step()`。
- 日志输出每组参数数量和 optimizer 类型。

对于 sparse reinit：

- 不要在 trainer 里硬编码 `torch.optim.Adagrad`。
- builder 或 helper 提供 `rebuild_sparse_optimizer(...)`。
- 只有 sparse optimizer 需要随 high-cardinality embedding reinit 重建；Muon / AdamW 参数组不需要重建。

### 6. CLI 参数

建议在 `parse_pcvr_train_args` 中新增：

```text
--optimizer legacy|adamw|muon_adamw
--dense_optimizer adamw|muon_adamw
--sparse_optimizer adagrad|none
--adamw_betas 0.9,0.98
--adamw_eps 1e-8
--adamw_weight_decay 0.01
--muon_lr
--muon_momentum
--muon_nesterov
--muon_ns_steps
--muon_weight_decay
--muon_backend torch
--muon_min_ndim 2
--muon_include_patterns
--muon_exclude_patterns
```

为了避免参数面一下子过宽，MVP 可以只实现：

- `--optimizer legacy|muon_adamw`
- `--adamw_weight_decay`
- `--muon_lr`
- `--muon_momentum`
- `--muon_ns_steps`
- `--muon_weight_decay`
- `--muon_min_ndim`

后续再补 include/exclude pattern。

### 7. 文档和实验落点

需要更新：

- `docs/experiments/baseline.md`：说明默认仍是 Adagrad + AdamW，Muon 需要显式开启。
- `docs/experiments/unirec.md`：把“2D 结构化权重走 Muon”从文档目标落到可执行配置上。
- 可选新增 `docs/guide/optimizer.md` 或在训练指南中加入 optimizer 参数表。

推荐先用 `config/baseline` 或 `config/unirec` 做最小 ablation：

- legacy：Adagrad + AdamW。
- muon_adamw：Adagrad + Muon + AdamW fallback。
- 记录参数分组数量、训练速度、显存、验证 AUC / logloss。

## 非目标

- 第一阶段不接 TorchRec `CombinedOptimizer` 或 fbgemm rowwise optimizer。
- 第一阶段不做 Triton/fused Muon kernel。
- 不改变模型架构和数据管道。
- 不改变评估、推理和在线 bundle 的运行入口。
- 不把 Muon 应用于 embedding、bias、LayerNorm/RMSNorm 一维权重。

## 实施清单

- [ ] 新增 `PCVROptimizerConfig` 和 `src/taac2026/infrastructure/pcvr/optimizers.py`。
- [ ] 实现 pure PyTorch `Muon` optimizer，支持 CPU 单元测试和 CUDA 训练。
- [ ] 实现参数分组：sparse embedding、Muon-eligible structured weights、AdamW fallback。
- [ ] 在 `parse_pcvr_train_args` 暴露 MVP optimizer 参数，并写入 `train_config` sidecar。
- [ ] 重构 `PCVRPointwiseTrainer`，使用 optimizer builder 返回的命名 optimizer 列表。
- [ ] 保留 `optimizer=legacy` 默认行为，并确保现有训练配置不变。
- [ ] 改造 high-cardinality sparse reinit，避免 trainer 继续硬编码 Adagrad。
- [ ] 训练日志输出每组 optimizer 的 tensor 数、参数量、lr、weight decay。
- [ ] 更新 baseline / unirec / training 文档。
- [ ] 做一次 smoke ablation，记录 legacy vs muon_adamw 的训练速度和验证指标。

## 测试要求

建议新增或更新：

- `tests/unit/test_pcvr_optimizers.py`
  - 参数分组不会重叠，且覆盖所有 `requires_grad=True` 参数。
  - embedding 参数进入 sparse optimizer，不进入 Muon。
  - `nn.Linear.weight` 进入 Muon；bias 和 norm 权重进入 AdamW fallback。
  - `optimizer=legacy` 构造结果与当前 Adagrad + AdamW 语义一致。
  - Muon 单步更新后参数有限值，shape 不变。
  - include/exclude pattern 如果实现，需要覆盖。
- `tests/unit/test_pcvr_trainer.py`
  - trainer 可以接受多个 optimizer 并按顺序 zero/step。
  - sparse reinit 后只重建 sparse optimizer，并恢复低基数 embedding state。
- `tests/unit/test_training_cli.py`
  - 新 optimizer CLI 参数能正确转发到 PCVR 训练参数。
- 可选 GPU 测试
  - `optimizer=muon_adamw` 在 CUDA 上跑一个最小 forward/backward/step smoke。

## 验收标准

- 默认训练命令行为不变：未显式传 `--optimizer` 时仍使用当前 Adagrad + AdamW 组合。
- 显式传 `--optimizer muon_adamw` 时，2D dense weight 进入 Muon，fallback 参数进入 AdamW，embedding 参数仍走 sparse optimizer。
- `train_config` sidecar 完整记录 optimizer 类型和关键超参，保证实验可复现。
- 日志能清楚打印每个 optimizer group 的参数数量，方便确认 Muon 没有误选 embedding 或 1D 参数。
- 单元测试覆盖参数分组、Muon step、legacy 兼容和 trainer 多 optimizer 调用。
- `uv run pytest tests/unit -q` 通过。
- 至少完成一次小样本 smoke，对比 legacy 与 `muon_adamw` 的训练可运行性。
